sklearn/linear_model/ridge.py

```diff
@@ -1015,48 +1246,152 @@ def _errors_and_values_helper(self, alpha, y, v, Q, QT_y):
+            self, alpha, y, sqrt_sw, X_mean, s, V, X):
+        """Compute dual coefficients and diagonal of (Identity - Hat_matrix)
+
+        Used when we have a decomposition of X^T.X
+        (n_features < n_samples and X is sparse), and not fitting an intercept.
+        """
+        n_samples, n_features = X.shape
+        w = 1 / (s + alpha)
+        A = (V * w).dot(V.T)
+        AXy = A.dot(safe_sparse_dot(X.T, y, dense_output=True))
+        y_hat = safe_sparse_dot(X, AXy, dense_output=True)
+        hat_diag = self._sparse_multidot_diag(X, A, X_mean, sqrt_sw)
+        if len(y.shape) != 1:
+            # handle case where y is 2-d
+            hat_diag = hat_diag[:, np.newaxis]
+        return (1 - hat_diag) / alpha, (y - y_hat) / alpha
+
+    def _errors_and_values_covariance_sparse_intercept(
+            self, alpha, y, sqrt_sw, X_mean, s, V, X):
+        """Compute dual coefficients and diagonal of (Identity - Hat_matrix)
+
+        Used when we have a decomposition of X^T.X
+        (n_features < n_samples and X is sparse),
+        and we are fitting an intercept.
+        """
+        n_samples, n_features = X.shape
+        # the vector [0, 0, ..., 0, 1]
+        # is the eigenvector of X^TX which
+        # corresponds to the intercept; we cancel the regularization on
+        # this dimension. the corresponding eigenvalue is
+        # sum(sample_weight), e.g. n when uniform sample weights.
+        intercept_sv = np.zeros(V.shape[0])
+        intercept_sv[-1] = 1
+        intercept_dim = _find_smallest_angle(intercept_sv, V)
+        w = 1 / (s + alpha)
+        w[intercept_dim] = 1 / s[intercept_dim]
+        A = (V * w).dot(V.T)
+        # add a column to X containing the square roots of sample weights
+        X_op = _X_operator(X, X_mean, sqrt_sw)
+        AXy = A.dot(X_op.T.dot(y))
+        y_hat = X_op.dot(AXy)
+        hat_diag = self._sparse_multidot_diag(X, A, X_mean, sqrt_sw)
+        # return (1 - hat_diag), (y - y_hat)
+        if len(y.shape) != 1:
+            # handle case where y is 2-d
+            hat_diag = hat_diag[:, np.newaxis]
+        return (1 - hat_diag) / alpha, (y - y_hat) / alpha
+
+    def _errors_and_values_covariance_sparse(
+            self, alpha, y, sqrt_sw, X_mean, s, V, X):
+        """Compute dual coefficients and diagonal of (Identity - Hat_matrix)
+
+        Used when we have a decomposition of X^T.X
+        (n_features < n_samples and X is sparse).
+        """
+        if self.fit_intercept:
+            return self._errors_and_values_covariance_sparse_intercept(
+                alpha, y, sqrt_sw, X_mean, s, V, X)
+        return self._errors_and_values_covariance_sparse_no_intercept(
+            alpha, y, sqrt_sw, X_mean, s, V, X)
+
+    def _decompose_covariance_dense(self, X, y, sqrt_sw):
+        # X already centered
+        X_mean = np.zeros(X.shape[1], dtype=X.dtype)
+        if self.fit_intercept:
+            intercept = sqrt_sw[:, None]
```

Comment:
Because this is not the actual intercept value but a pseudo column that serves to emulate fitting an intercept, maybe it could be renamed to `intercept_column`.